File: iosApp/iosApp/Modules/NFTsDashboard/NFTsDashboardWireframeFactory.swift
// Created by web3d4v on 24/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - NFTsDashboardWireframeFactory

protocol NFTsDashboardWireframeFactory {
    func make(_ parent: UIViewController?) -> NFTsDashboardWireframe
}

// MARK: - DefaultNFTsDashboardWireframeFactory

final class DefaultNFTsDashboardWireframeFactory {
    private let nftsCollectionWireframeFactory: NFTsCollectionWireframeFactory
    private let nftDetailWireframeFactory: NFTDetailWireframeFactory
    private let nftsService: NFTsService
    private let networksService: NetworksService
    private let settingsService: SettingsService
    private let mailService: MailService

    init(
        nftsCollectionWireframeFactory: NFTsCollectionWireframeFactory,
        nftDetailWireframeFactory: NFTDetailWireframeFactory,
        nftsService: NFTsService,
        networksService: NetworksService,
        settingsService: SettingsService,
        mailService: MailService
    ) {
        self.nftsCollectionWireframeFactory = nftsCollectionWireframeFactory
        self.nftDetailWireframeFactory = nftDetailWireframeFactory
        self.nftsService = nftsService
        self.networksService = networksService
        self.settingsService = settingsService
        self.mailService = mailService
    }
}

extension DefaultNFTsDashboardWireframeFactory: NFTsDashboardWireframeFactory {

    func make(_ parent: UIViewController?) -> NFTsDashboardWireframe {
        DefaultNFTsDashboardWireframe(
            parent,
            nftsCollectionWireframeFactory: nftsCollectionWireframeFactory,
            nftDetailWireframeFactory: nftDetailWireframeFactory,
            nftsService: nftsService,
            networksService: networksService,
            settingsService: settingsService,
            mailService: mailService
        )
    }
}

// MARK: - Assembler

final class NFTsDashboardWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> NFTsDashboardWireframeFactory in
            DefaultNFTsDashboardWireframeFactory(
                nftsCollectionWireframeFactory: resolver.resolve(),
                nftDetailWireframeFactory: resolver.resolve(),
                nftsService: resolver.resolve(),
                networksService: resolver.resolve(),
                settingsService: resolver.resolve(),
                mailService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftsdashboard/NFTsDashboardWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftsdashboard

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail.NFTDetailWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftscollection.DefaultNFTsCollectionWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftscollection.NFTsCollectionWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframe
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService
import smartadapter.internal.extension.name

interface NFTsDashboardWireframeFactory {
    fun make(parent: Fragment?): NFTsDashboardWireframe
}

class DefaultNFTsDashboardWireframeFactory(
    private val networksService: NetworksService,
    private val nftsService: NFTsService,
    private val nftsCollectionWireframeFactory: DefaultNFTsCollectionWireframeFactory,
    private val nftDetailWireframeFactory: NFTDetailWireframeFactory,
): NFTsDashboardWireframeFactory {

    override fun make(parent: Fragment?): NFTsDashboardWireframe = DefaultNFTsDashboardWireframe(
        parent?.let { WeakRef(it) },
        networksService,
        nftsService,
        nftsCollectionWireframeFactory,
        nftDetailWireframeFactory,
    )
}

class NFTsDashboardWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {
        to.register(NFTsDashboardWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultNFTsDashboardWireframeFactory(
                it.resolve(NetworksService::class.name),
                it.resolve(NFTsService::class.name),
                it.resolve(NFTsCollectionWireframeFactory::class.name),
                it.resolve(NFTDetailWireframeFactory::class.name),
            )
        }
    }
}
